### ISSUE TITLE
[xgboost4j-spark] Checkpointing performance issue

### ISSUE THREAD

**Issue by alois-bissuel** [dmlc/xgboost#3946]
I just noticed a weird performance issue when using checkpoints in the Spark wrapper for XGBoost.
Before the first checkpoint, the learning time is constant per tree, but after the first checkpoint, the time needed to learn a new tree is much higher, and it grows with the number of trees. 

Here is the parameter map used by XGBoost:
`Map(alpha -> 0.0, min_child_weight -> 1.0, sample_type -> uniform, base_score -> 0.5, colsample_bylevel -> 1.0, grow_policy -> depthwise, skip_drop -> 0.0, lambda_bias -> 0.0, silent -> 0, scale_pos_weight -> 1.0, seed -> 0, features_col -> features, num_early_stopping_rounds -> 0, label_col -> label, num_workers -> 5, subsample -> 1.0, lambda -> 1.0, max_depth -> 5, probability_col -> probability, raw_prediction_col -> rawPrediction, tree_limit -> 0, custom_eval -> null, rate_drop -> 0.0, max_bin -> 16, train_test_ratio -> 1.0, use_external_memory -> false, objective -> binary:logistic, eval_metric -> logloss, num_round -> 500, timeout_request_workers -> 1800000, missing -> NaN, checkpoint_path -> viewfs://root/user/XXX/YYYY/checkpoint, tracker_conf -> TrackerConf(0,python), tree_method -> auto, max_delta_step -> 0.0, eta -> 0.3, colsample_bytree -> 1.0, normalize_type -> tree, custom_obj -> null, gamma -> 0.0, sketch_eps -> 0.03, nthread -> 4, prediction_col -> prediction, checkpoint_interval -> 100)`

Attached is a small graph, where this issue is shown on the same dataset, with different checkpointing number. Please disregard the discrepancy between the two run, as they might not have been launched with the same multithreading parameters.

<img width="877" alt="image" src="https://user-images.githubusercontent.com/41194547/49159616-35fe5500-f325-11e8-94ad-fa3b8950da2d.png">

A quick profiling (using async-profiler) before and after the checkpoint might give a hint to the problem. Before the checkpoint, I see many calls to `tree::CQHistMaker` or `tree::GlobalProposalHistMaker`, but after the checkpoint point, the C++ calls are all to `CPUPredictor::PredLoopSpecialize`. I do not know whether this means that the method used for learning is not the same after the checkpoint, or that the time needed to evaluate the tree is so long that it the profiler only sees it. I could upload some flamegraphs, but I would need first to make sure they are actually representative of what is going on.

**Comment by chenqin** [comment]
IMO, there were unnecessary reshuffle after each checkpoint. Is there any info around memory utilization?

**Comment by CodingCat** [comment]
@cq not true, extra shuffle happens for every 100 iterations in this case..

**Comment by alois-bissuel** [comment]
@chenqin and @CodingCat , do you need any other "benchmark" ?

**Comment by AnthonyTruchet** [comment]
Hello, has this question raised any interest? we are willing to investigate and contribute and for that wanted to check first that no update / fix existed... Is anyone watching/involved?

**Comment by hcho3** [comment]
@CodingCat Is anyone in your org looking at this issue?

**Comment by CodingCat** [comment]
no, please go ahead, we do not have bandwidth to take care of it for now

**Comment by bourbaki** [comment]
I have started investigating the issue. And it looks like I found one possible cause of it. The prediction caches are not initialized after the first checkpoint (`gbm_.predictor_.cache_`). And if the caches have not been initialized they will not be filled. So every iteration all predictions for every previous tree are recomputed. We are losing caches when propagating the booster to the driver. There are two possible solutions:
1. Reinitialize the caches after starting from checkpoint. It will solve the part of the problem. The cost is that we need to recompute caches on the first iteration after every checkpoint.
2. Propagate the caches with a booster. I think there is some kind of issue with serialization of C++ object.

**Comment by CodingCat** [comment]
@bourbaki thanks for the analysis and I agree with it 

regarding how to resolve it, I think those two solutions have their own problems as you said

1. the first iteration right after the checkpoint would be super slow, and it becomes slower and slower with the growth of trees

2. the problem is we have to change the serialization format of booster which may hurt the other bindings of xgboost, and furthermore, the serialized booster would be in a huge size if the prediction results are happen to be very large

the ideal fix should be more fundamental...I got some ideas and am about to start my work...but I just noticed you added `I am preparing a pull request` after I read your comments last time

please let me know your proposal before you submit the code, and I am actually in a hurry to get this fix due to an internal request

**Comment by CodingCat** [comment]
@bourbaki if you are dedicated to this, I can assign this to you

**Comment by bourbaki** [comment]
@CodingCat Yes. I am dedicated to this. What approach would you suggest?

**Comment by CodingCat** [comment]
@bourbaki sorry for the late reply, really busy in these days, my suggestion would be 

1. since the checkpoint is a functionality built in xgboost4j-spark layer, we can just pursue a solution in this level, 

2. the point is to get cache initialized with a booster from checkpoint, I would suggest we 

2.1 save the prediction results with Spark

2.2 expose an API to directly call [Init Function in Predictor](https://github.com/dmlc/xgboost/blob/master/src/predictor/predictor.cc#L11-L17) 

2.3 when starting training from a checkpoint, we should load back data (maybe have to through HDFS API directly) saved in 2.1, and use a resembled version of [PredLoopSpecalize](https://github.com/dmlc/xgboost/blob/master/src/predictor/cpu_predictor.cc#L44-L94) to update the cache with the loaded Dataset


the most important thing we need to be careful is to guarantee the deterministic partitioning

**Comment by bourbaki** [comment]
It sounds good. But there's one thing that bothers me with the current design of `xgboost4k-spark`.
At every checkpoint we collect booster data from executors to a driver to save it. But then to start the next iteration instead of reusing the current state of executors we copy a booster from the driver to each executor. These copies are redundant in the case when there are no failures in the training proces and we just want to get the current state of a booster without interrupting the training process. It would be cool to have something like this:
```
val boosters1: RDD[Booster] = train(..., iterations=100).cache
val boosterToSave = boosters1.first
boosterToSave.save(...)
val boosters2 = continueTrain(boosters1, iterations=100).cache
....
```

**Comment by CodingCat** [comment]
> It sounds good. But there's one thing that bothers me with the current design of `xgboost4k-spark`.
> At every checkpoint we collect booster data from executors to a driver to save it. But then to start the next iteration instead of reusing the current state of executors we copy a booster from the driver to each executor. These copies are redundant in the case when there are no failures in the training proces and we just want to get the current state of a booster without interrupting the training process. It would be cool to have something like this:
> 
> ```
> val boosters1: RDD[Booster] = train(..., iterations=100).cache
> val boosterToSave = boosters1.first
> boosterToSave.save(...)
> val boosters2 = continueTrain(boosters1, iterations=100).cache
> ....
> ```
> In this case when training goes according to the plan we only copy booster to driver and then restart the training from the cached state.

yes, in that way we need some way to pass the message from driver to executor saying "please move on"...leading us to more complexities from the message system for training...

if we want to pursue that approach, we need to extend rabit layer to allow tracker and worker to communicate more than "start/stop/I died/please reborn", etc.

**Comment by CodingCat** [comment]
but still we need to save the prediction results to allow the training recover from a checkpoint..which does not reduce the complexity I talked about earlier

**Comment by CodingCat** [comment]
@bourbaki are you working on this?

**Comment by CodingCat** [comment]
ping @bourbaki

**Comment by bourbaki** [comment]
Sorry for the late reply. I have been busy for the last couple of weeks. I started to work on it. I am planning to spend some time on this week.

**Comment by CodingCat** [comment]
@bourbaki are you still working on this? if you are too busy, I can take over it

**Comment by bourbaki** [comment]
@CodingCat Sorry, I am too busy right now. You can take over it.

**Comment by trams** [comment]
Hello @CodingCat. We allocated significant time to fix checkpointing performance issue in XGBoost4J-Spark (it is a blocker for us) in July-August. So I am wandering what is a current status of your work and whether you prepare any pull requests now or I could take this over.

Thank you

**Comment by CodingCat** [comment]
I am currently working on it

**Comment by trams** [comment]
@CodingCat could you please share the status of your work and confirm the approach you decided to take?

This issue blocks us (Criteo). We allocated resources to fix it. Let us know how we can contribute or how you can hand over if you have higher priorities.

**Comment by CodingCat** [comment]
it’s still WIP in my side, and I will share the design in the format of a RFC within this month (actually I have shared some in this thread) and code later ,

**Comment by CodingCat** [comment]
update: 

basically, I have run the prototype fix with our production data and everything looks fine. Now, it comes to the cost of fix. The fix under test goes to the path that the prediction cache is rebuilt for each checkpoint interval, however, it still slows down first iteration of each interval. However, a more fundamental fix requires API changes/adding, and the added API might be only consumed by jvm package, we need to evaluate whether it is worth that

let me start a new design discussion thread on this

**Comment by trams** [comment]
Thank you for the update. I am glad you have a fix and from your description I think I have a similar fix being tested on production dataset too.

I am looking forward for you sharing notes|patches so we can compare our notes.

By the way, my fix is published as a branch in our fork. I can link it here if you are interested.
I did not submit a pull request cause it is not yet ready for it: there are some "unit tests" to show the effect of checkpointing and some not very clean logging.

**Comment by trams** [comment]
@CodingCat do you have any updates on your side?

I want to prepare a RFC with my idea for potential fixes. What do you think?

**Comment by hcho3** [comment]
Can this be closed, now that #5082 is merged?
